supermarq-benchmarks/supermarq/qcvv/base_experiment.py

```diff
@@ -0,0 +1,462 @@
+            RuntimeError: If no samples are available.
+        """
+        if self._samples is None:
+            raise RuntimeError("No samples to retrieve. The experiment has not been run.")
+
+        return self._samples
+
+    @property
+    def raw_data(self) -> pd.DataFrame:
+        """The data from the most recently run experiment.
+
+        Raises:
+            RuntimeError: If no results are available.
+        """
+        if self._raw_data is None:
+            raise RuntimeError("No data to retrieve. The experiment has not been run.")
+
+        return self._raw_data
+
+    @property
+    def num_qubits(self) -> int:
+        """Returns:
+        The number of qubits used in the experiment
+        """
+        return len(self.qubits)
+
+    def run(
+        self,
+        num_circuits: int,
+        layers: Iterable[int],
+        target: cirq.SimulatorBase | str | None = None,  # type: ignore [type-arg]
+        shots: int = 10_000,
+        method: str | None = None,
+        target_options: dict[str, Any] | None = None,
+    ) -> None:
+        """Run the benchmarking experiment and analyse the results.
+
+        Args:
+            num_circuits: Number of circuits to run.
+            layers: An iterable of the different layer depths to use during the experiment.
+            target: Either a local :class:`~cirq.SimulatorBase` to use or the name of a Superstaq
+                target. If None then a local simulator is used. Defaults to None.
+            shots: The number of shots to sample. Defaults to 10,000.
+            method: Optional method to use on the Superstaq device. Defaults to None corresponding
+                to normal running.
+            target_options: Optional configuration dictionary passed when submitting the job.
+        """
+        if self._samples is not None:
+            warnings.warn("Existing results will be overwritten.")
+
+        if any(depth <= 0 for depth in layers):
+            raise ValueError("The `layers` iterator can only include positive values.")
+
+        self._samples = self.build_circuits(num_circuits, layers)
+        self._clean_circuits()
+
+        if target is None:
+            target = cirq.Simulator()
+
+        if isinstance(target, str):
+            self.submit_ss_jobs(target, shots, method, **(target_options or {}))
+        else:
+            self.sample_circuits_with_simulator(target, shots)
+
+    def _clean_circuits(self) -> None:
+        """Removes any terminal measurements that have been added to the circuit and replaces
+        them with a single measurement of the whole system in the computational basis
+        """
+        for sample in self.samples:
+            if sample.circuit.has_measurements():
+                sample.circuit = cirq.drop_terminal_measurements(sample.circuit)
+            # Add measurement of qubit state in computational basis.
+            sample.circuit += cirq.measure(sample.circuit.all_qubits())
+
+    def submit_ss_jobs(
```

Comment:
nit: i don't think we need the `ss` here - it should be clear enough that that's how we're submitting them
```suggestion
    def submit_jobs(
```
(but if we do want to keep it, imo we should use `_superstaq_` instead of `_ss_`)